src/ast/effects.h

```diff
@@ -164,10 +169,24 @@ struct EffectAnalyzer : public PostWalker<EffectAnalyzer> {
   }
   void visitLoad(Load *curr) {
     readsMemory = true;
+    isAtomic = curr->isAtomic;
```

Comment:
these flags are for the entire expression being scanned, including all its children. so `readsMemory` means that in all our scanning, we found a read of memory, so the whole thing reads memory (somewhere inside it). so if we scan two loads, one atomic and the second non-atomic, the current code would mark them all as not being atomic.